rfcs/0004-policy-sync-v1.md

```diff
@@ -0,0 +1,182 @@
+
+# Summary
+[summary]: #summary
+
+The ability for the Multicluster Gateway Controller to sync policies defined in
+the hub cluster downstream to the spoke clusters, therefore allowing all policies
+to be defined in the same place. These policies will be reconciled by the downstream
+policy controller(s).
+
+# Nomenclature
+
+* Policy: When refering to a Policy, this document is refering to a Gateway API
+  policy as defined in the Policy Attachment Model. The Multicluster Gateway Controller
+  relies on [OCM](https://open-cluster-management.io/) as a Multicluster solution, which defines its own unrelated
+  set of Policies and Policy Framework. Unless explicitely mentioned, this document
+  refers to Policies as Gateway API Policies.
+
+* Policy overriding: The concept of policy overriding is mentioned in this document. It refers to the proposed ability of the downstream Gateway implementation to prioritise downstream Policies against synced Policies in case of conflicts.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, Kuadrant's support for the Policy Attachment Model can be divided in
+two categories:
+* Policies targeting the Multicluster Gateway, defined in the hub cluster and
+  reconciled by the Multicluster Gateway Controller
+* Policies targeting the downstream Gateway, defined in the spoke clusters and
+  reconciled by the downstream Gateway controllers.
+
+In a realistic multicluster scenario where multiple spoke clusters are present, the management of these policies can become tedious and error-prone, as policies have
+to be defined in the hub cluster, as well as replicated in the multiple spoke clusters.
+
+As Kuadrant users:
+* Gateway-admin has a set of homogeneous clusters and needs to apply per cluster rate limits across the entire set.
+* Platform-admin with a set of clusters with rate limits applied needs to change rate limit for one particular cluster. 
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The policy sync feature will allow a gateway-admin to configure, via GatewayClass
+parameters, a set of Policy GVRs to be synced by the Multicluster Gateway Controller.
+
+The `policiesToSync` field in the parameters defines those GVRs. For example, in
+order to configure the controller to sync AuthPolicies:
+
+```json
+"policiesToSync": [
+  {
+    "group": "kuadrant.io",
+    "version": "v1beta1",
+    "resource": "authpolicies" 
+  }
+]
+```
+
+The support for resources that the controller can sync is limited by the following:
+* The controller ServiceAccount must have permission to watch, list, and get the
+  resource to be synced
+* The resource must implement the Policy schema:
+    * Have a `.spec.targetRef` field
+
+When a Policy is configured to be synced in a GatewayClass, the Multicluster
+Gateway Controller starts watching events on the resources, and propagates changes
+by placing the policy in the spoke clusters, with the following mutations:
```

Comment:
Will there be any annotations added to the original policy as well? I think it would be useful if I could trace the spoke policies from some annotation, similar to how Gateway `kuadrant.io/gateway-clusters` works

Reply:
That's a good idea. Would it be an annotation to reference the downstream policies or simply to inform that it has been synced?

Reply:
I would say the former, a reference would be nice

Reply:
Yes, I agree. I'll add it to the document